File: scripts/_skill_tools.py
#!/usr/bin/env python3
"""Shared, dependency-free helpers for repository-level skill tooling."""

from __future__ import annotations

import json
import os
import re
import shutil
import subprocess
import sys
import uuid
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path


SKILL_NAME_RE = re.compile(r"^[a-z0-9]+(?:-[a-z0-9]+)*$")
MARKDOWN_LINK_RE = re.compile(r"(?<!!)\[[^\]]+\]\(([^)]+)\)")
IGNORED_TOP_LEVEL_DIRS = {"evals"}
IGNORED_DIR_NAMES = {"__pycache__", ".git", ".pytest_cache"}
IGNORED_FILE_NAMES = {".DS_Store"}
IGNORED_FILE_SUFFIXES = {".pyc", ".pyo"}


class SkillToolError(RuntimeError):
    """A user-actionable repository tooling error."""


@dataclass(frozen=True)
class SkillMetadata:
    name: str
    description: str


@dataclass(frozen=True)
class InstallResult:
    action: str
    source: Path
    destination: Path
    backup: Path | None = None


def repository_root() -> Path:
    return Path(__file__).resolve().parent.parent


def _frontmatter_text(skill_md: Path) -> str:
    text = skill_md.read_text(encoding="utf-8")
    lines = text.splitlines()
    if not lines or lines[0].strip() != "---":
        raise SkillToolError(f"{skill_md} must start with YAML frontmatter")
    try:
        end = next(index for index, line in enumerate(lines[1:], start=1) if line.strip() == "---")
    except StopIteration as exc:
        raise SkillToolError(f"{skill_md} has unterminated YAML frontmatter") from exc
    return "\n".join(lines[1:end])


def _unquote(value: str) -> str:
    value = value.strip()
    if len(value) >= 2 and value[0] == value[-1] and value[0] in {'"', "'"}:
        return value[1:-1]
    return value


def read_skill_metadata(skill_dir: Path) -> SkillMetadata:
    skill_md = skill_dir / "SKILL.md"
    if not skill_md.is_file():
        raise SkillToolError(f"missing SKILL.md: {skill_dir}")

    frontmatter = _frontmatter_text(skill_md)
    name_match = re.search(r"(?m)^name:\s*(.+?)\s*$", frontmatter)
    if not name_match:
        raise SkillToolError(f"{skill_md} is missing frontmatter field 'name'")
    name = _unquote(name_match.group(1))

    frontmatter_lines = frontmatter.splitlines()
    description_index = next(
        (index for index, line in enumerate(frontmatter_lines) if line.startswith("description:")),
        None,
    )
    if description_index is None:
        raise SkillToolError(f"{skill_md} is missing frontmatter field 'description'")
    description_value = frontmatter_lines[description_index].split(":", 1)[1].strip()
    if description_value in {">", ">-", "|", "|-"}:
        parts: list[str] = []
        for line in frontmatter_lines[description_index + 1 :]:
            if not line.startswith((" ", "\t")):
                break
            parts.append(line.strip())
        description = " ".join(part for part in parts if part)
    else:
        description = _unquote(description_value)

    return SkillMetadata(name=name, description=description)


def validate_skill(skill_dir: Path, expected_name: str | None = None) -> list[str]:
    skill_dir = skill_dir.resolve()
    issues: list[str] = []
    try:
        metadata = read_skill_metadata(skill_dir)
    except (OSError, SkillToolError) as exc:
        return [str(exc)]

    required_name = expected_name or skill_dir.name
    if expected_name is None and not SKILL_NAME_RE.fullmatch(skill_dir.name):
        issues.append(f"directory name is not lowercase hyphen-case: {skill_dir.name}")
    if metadata.name != required_name:
        issues.append(
            f"frontmatter name {metadata.name!r} does not match expected name {required_name!r}"
        )
    if not SKILL_NAME_RE.fullmatch(metadata.name):
        issues.append(f"invalid skill name: {metadata.name!r}")
    if not metadata.description:
        issues.append("description must not be empty")
    if len(metadata.description) > 1024:
        issues.append(f"description exceeds 1024 characters: {len(metadata.description)}")

    for json_file in sorted(skill_dir.rglob("*.json")):
        if any(part in IGNORED_DIR_NAMES for part in json_file.relative_to(skill_dir).parts):
            continue
        try:
            json.loads(json_file.read_text(encoding="utf-8"))
        except (OSError, json.JSONDecodeError) as exc:
            issues.append(f"invalid JSON {json_file.relative_to(skill_dir)}: {exc}")

    for markdown_file in sorted(skill_dir.rglob("*.md")):
        if any(part in IGNORED_DIR_NAMES for part in markdown_file.relative_to(skill_dir).parts):
            continue
        try:
            markdown = markdown_file.read_text(encoding="utf-8")
        except OSError as exc:
            issues.append(f"cannot read {markdown_file.relative_to(skill_dir)}: {exc}")
            continue
        for raw_target in MARKDOWN_LINK_RE.findall(markdown):
            target = raw_target.strip().strip("<>").split("#", 1)[0]
            if not target or "://" in target or target.startswith(("mailto:", "/")):
                continue
            linked = (markdown_file.parent / target).resolve()
            if not linked.exists():
                issues.append(
                    f"broken link in {markdown_file.relative_to(skill_dir)}: {raw_target}"
                )

    return issues


def require_valid_skill(skill_dir: Path, expected_name: str | None = None) -> SkillMetadata:
    issues = validate_skill(skill_dir, expected_name=expected_name)
    if issues:
        formatted = "\n  - ".join(issues)
        raise SkillToolError(f"skill validation failed:\n  - {formatted}")
    return read_skill_metadata(skill_dir)


def resolve_source(skill: str, repo_root: Path | None = None) -> tuple[Path, SkillMetadata]:
    repo_root = (repo_root or repository_root()).resolve()
    skills_root = (repo_root / "skills").resolve()
    requested = Path(skill).expanduser()
    if requested.is_absolute() or requested.exists() or "/" in skill or "\\" in skill:
        source = requested.resolve()
    else:
        if not SKILL_NAME_RE.fullmatch(skill):
            raise SkillToolError(f"invalid skill name: {skill!r}")
        source = (skills_root / skill).resolve()

    try:
        source.relative_to(skills_root)
    except ValueError as exc:
        raise SkillToolError(f"skill source must be inside {skills_root}: {source}") from exc
    if not source.is_dir():
        raise SkillToolError(f"skill source directory does not exist: {source}")
    metadata = require_valid_skill(source)
    return source, metadata


def resolve_target(agent: str | None, target_dir: str | None) -> Path:
    if target_dir:
        expanded = os.path.expandvars(target_dir)
        return Path(expanded).expanduser().resolve()

    profile = agent or "codex"
    if profile == "codex":
        config_root = Path(os.environ.get("CODEX_HOME", Path.home() / ".codex"))
    elif profile == "pi":
        config_root = Path(
            os.environ.get("PI_CODING_AGENT_DIR", Path.home() / ".pi" / "agent")
        )
    else:
        raise SkillToolError(f"unsupported agent profile: {profile}")
    return config_root.expanduser().resolve() / "skills"


def _destination_exists(path: Path) -> bool:
    return path.exists() or path.is_symlink()


def _same_symlink(destination: Path, source: Path) -> bool:
    return destination.is_symlink() and destination.resolve(strict=False) == source.resolve()


def _copy_ignore(source: Path):
    source = source.resolve()

    def ignore(directory: str, names: list[str]) -> set[str]:
        directory_path = Path(directory).resolve()
        relative = directory_path.relative_to(source)
        ignored: set[str] = set()
        for name in names:
            if name in IGNORED_DIR_NAMES or name in IGNORED_FILE_NAMES:
                ignored.add(name)
                continue
            if Path(name).suffix in IGNORED_FILE_SUFFIXES:
                ignored.add(name)
                continue
            if relative == Path(".") and name in IGNORED_TOP_LEVEL_DIRS:
                ignored.add(name)
        return ignored

    return ignore


def _backup_path(target_root: Path, skill_name: str) -> Path:
    stamp = datetime.now(timezone.utc).strftime("%Y%m%dT%H%M%SZ")
    base = target_root.parent / ".skill-install-backups" / target_root.name / stamp
    candidate = base / skill_name
    counter = 2
    while _destination_exists(candidate):
        candidate = base / f"{skill_name}-{counter}"
        counter += 1
    return candidate


def install_skill(
    source: Path,
    metadata: SkillMetadata,
    target_root: Path,
    mode: str = "link",
    replace: bool = False,
    dry_run: bool = False,
) -> InstallResult:
    source = source.resolve()
    target_root = target_root.expanduser().resolve()
    destination = target_root / metadata.name

    if destination == source:
        raise SkillToolError("installation destination is the source directory itself")
    if source in target_root.parents:
        raise SkillToolError("installation target cannot be inside the skill source")
    if mode not in {"link", "copy"}:
        raise SkillToolError(f"unsupported install mode: {mode}")

    occupied = _destination_exists(destination)
    if occupied and mode == "link" and _same_symlink(destination, source):
        return InstallResult("already-installed", source, destination)
    if occupied and not replace:
        raise SkillToolError(
            f"destination already exists: {destination}; rerun with --replace to back it up"
        )

    backup = _backup_path(target_root, metadata.name) if occupied else None
    if dry_run:
        return InstallResult("would-replace" if occupied else "would-install", source, destination, backup)

    target_root.mkdir(parents=True, exist_ok=True)
    staging = target_root / f".{metadata.name}.install-{uuid.uuid4().hex}"
    try:
        if mode == "link":
            staging.symlink_to(source, target_is_directory=True)
        else:
            shutil.copytree(source, staging, ignore=_copy_ignore(source))
            require_valid_skill(staging, expected_name=metadata.name)

        if occupied and backup:
            backup.parent.mkdir(parents=True, exist_ok=True)
            destination.rename(backup)
        try:
            staging.rename(destination)
        except Exception:
            if backup and _destination_exists(backup) and not _destination_exists(destination):
                backup.rename(destination)
            raise
    finally:
        if staging.is_symlink():
            staging.unlink()
        elif staging.exists():
            shutil.rmtree(staging)

    return InstallResult("replaced" if occupied else "installed", source, destination, backup)


def uninstall_skill(skill_name: str, target_root: Path, apply: bool = False) -> InstallResult:
    if not SKILL_NAME_RE.fullmatch(skill_name):
        raise SkillToolError(f"invalid skill name: {skill_name!r}")
    target_root = target_root.expanduser().resolve()
    destination = target_root / skill_name
    if not _destination_exists(destination):
        return InstallResult("not-installed", destination, destination)

    if not destination.is_symlink():
        metadata = require_valid_skill(destination)
        if metadata.name != skill_name:
            raise SkillToolError(
                f"refusing to remove destination with mismatched skill name: {metadata.name}"
            )

    if not apply:
        return InstallResult("would-uninstall", destination, destination)
    if destination.is_symlink() or destination.is_file():
        destination.unlink()
    else:
        shutil.rmtree(destination)
    return InstallResult("uninstalled", destination, destination)


def discover_skills(repo_root: Path | None = None) -> list[Path]:
    root = (repo_root or repository_root()).resolve() / "skills"
    if not root.is_dir():
        raise SkillToolError(f"skills directory does not exist: {root}")
    return sorted(path.parent for path in root.glob("*/SKILL.md"))


def run_python_tests(skill_dir: Path) -> subprocess.CompletedProcess[str] | None:
    tests_dir = skill_dir / "tests"
    if not tests_dir.is_dir() or not any(tests_dir.glob("test*.py")):
        return None
    return subprocess.run(
        [sys.executable, "-m", "unittest", "discover", "-s", str(tests_dir), "-v"],
        cwd=repository_root(),
        capture_output=True,
        text=True,
        check=False,
    )

File: scripts/skill-tools.ts
/** Dependency-free repository installation and validation. */
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { randomUUID } from "node:crypto";
import { spawnSync } from "node:child_process";

export const repositoryRoot = fileURLToPath(new URL("../", import.meta.url));
const namePattern = /^[a-z0-9]+(?:-[a-z0-9]+)*$/;
const ignoredDirs = new Set([
  "__pycache__",
  ".git",
  ".pytest_cache",
  "node_modules",
]);
export class SkillToolError extends Error {}
export interface SkillMetadata {
  name: string;
  description: string;
}
export interface InstallResult {
  action: string;
  source: string;
  destination: string;
  backup?: string;
}
export function exists(file: string): boolean {
  try {
    fs.lstatSync(file);
    return true;
  } catch (error) {
    if ((error as NodeJS.ErrnoException).code === "ENOENT") return false;
    throw error;
  }
}
/** Like Path.resolve(strict=False): resolve existing symlinks and missing tails. */
export function resolvePath(file: string, seen = new Set<string>()): string {
  const absolute = path.isAbsolute(file)
    ? file
    : process.cwd() + path.sep + file;
  const { root } = path.parse(absolute);
  let resolved = root;
  for (const part of absolute
    .slice(root.length)
    .split(path.sep)
    .filter(Boolean)) {
    if (part === ".") continue;
    if (part === "..") {
      resolved = path.dirname(resolved);
      continue;
    }
    resolved = path.join(resolved, part);
    if (exists(resolved) && fs.lstatSync(resolved).isSymbolicLink()) {
      if (seen.has(resolved))
        throw new SkillToolError(`Symlink cycle: ${resolved}`);
      const next = new Set(seen).add(resolved);
      const target = fs.readlinkSync(resolved);
      resolved = resolvePath(
        path.isAbsolute(target)
          ? target
          : path.dirname(resolved) + path.sep + target,
        next,
      );
    }
  }
  return resolved;
}
function expandUser(value: string): string {
  return value === "~"
    ? os.homedir()
    : value.startsWith("~/")
      ? path.join(os.homedir(), value.slice(2))
      : value;
}
export function readSkillMetadata(skill: string): SkillMetadata {
  const file = path.join(skill, "SKILL.md");
  if (!fs.existsSync(file) || !fs.statSync(file).isFile())
    throw new SkillToolError(`missing SKILL.md: ${skill}`);
  const lines = fs.readFileSync(file, "utf8").split(/\r?\n/);
  if (lines[0]?.trim() !== "---")
    throw new SkillToolError(`${file} must start with YAML frontmatter`);
  const end = lines.findIndex((line, i) => i > 0 && line.trim() === "---");
  if (end < 0)
    throw new SkillToolError(`${file} has unterminated YAML frontmatter`);
  const head = lines.slice(1, end);
  const unquote = (s: string) => s.trim().replace(/^(['"])(.*)\1$/, "$2");
  const name = head.join("\n").match(/^name:\s*(.+?)\s*$/m)?.[1];
  if (!name)
    throw new SkillToolError(`${file} is missing frontmatter field 'name'`);
  const i = head.findIndex((line) => line.startsWith("description:"));
  if (i < 0)
    throw new SkillToolError(
      `${file} is missing frontmatter field 'description'`,
    );
  let description = head[i].slice("description:".length).trim();
  if ([">", ">-", "|", "|-"].includes(description)) {
    const parts: string[] = [];
    for (const line of head.slice(i + 1)) {
      if (!/^[ \t]/.test(line)) break;
      if (line.trim()) parts.push(line.trim());
    }
    description = parts.join(" ");
  } else description = unquote(description);
  return { name: unquote(name), description };
}
function filesBelow(dir: string): string[] {
  return fs
    .readdirSync(dir, { withFileTypes: true })
    .sort((a, b) => a.name.localeCompare(b.name))
    .flatMap((entry) => {
      const file = path.join(dir, entry.name);
      if (entry.isDirectory())
        return ignoredDirs.has(entry.name) ? [] : filesBelow(file);
      return [file];
    });
}
export function validateSkill(
  directory: string,
  expectedName?: string,
): string[] {
  const skill = resolvePath(directory);
  const issues: string[] = [];
  let metadata: SkillMetadata;
  try {
    metadata = readSkillMetadata(skill);
  } catch (error) {
    return [String((error as Error).message)];
  }
  const required = expectedName ?? path.basename(skill);
  if (expectedName === undefined && !namePattern.test(path.basename(skill)))
    issues.push(
      `directory name is not lowercase hyphen-case: ${path.basename(skill)}`,
    );
  if (metadata.name !== required)
    issues.push(
      `frontmatter name '${metadata.name}' does not match expected name '${required}'`,
    );
  if (!namePattern.test(metadata.name))
    issues.push(`invalid skill name: ${metadata.name}`);
  if (!metadata.description) issues.push("description must not be empty");
  if ([...metadata.description].length > 1024)
    issues.push(
      `description exceeds 1024 characters: ${[...metadata.description].length}`,
    );
  for (const file of filesBelow(skill)) {
    if (file.endsWith(".json")) {
      try {
        JSON.parse(fs.readFileSync(file, "utf8"));
      } catch (error) {
        issues.push(
          `invalid JSON ${path.relative(skill, file)}: ${(error as Error).message}`,
        );
      }
    }
    if (file.endsWith(".md")) {
      let markdown: string;
      try {
        markdown = fs.readFileSync(file, "utf8");
      } catch (error) {
        issues.push(
          `cannot read ${path.relative(skill, file)}: ${(error as Error).message}`,
        );
        continue;
      }
      for (const match of markdown.matchAll(/(?<!!)\[[^\]]+\]\(([^)]+)\)/g)) {
        const target = match[1].trim().replace(/^<|>$/g, "").split("#", 1)[0];
        if (
          !target ||
          target.includes("://") ||
          target.startsWith("mailto:") ||
          target.startsWith("/")
        )
          continue;
        if (!fs.existsSync(path.resolve(path.dirname(file), target)))
          issues.push(
            `broken link in ${path.relative(skill, file)}: ${match[1]}`,
          );
      }
    }
  }
  return issues;
}
export function requireValidSkill(
  skill: string,
  expectedName?: string,
): SkillMetadata {
  const issues = validateSkill(skill, expectedName);
  if (issues.length)
    throw new SkillToolError(
      `skill validation failed:\n  - ${issues.join("\n  - ")}`,
    );
  return readSkillMetadata(skill);
}
const isWithin = (root: string, file: string) => {
  const relative = path.relative(root, file);
  return (
    relative === "" ||
    (!relative.startsWith(`..${path.sep}`) &&
      relative !== ".." &&
      !path.isAbsolute(relative))
  );
};
export function resolveSource(
  skill: string,
  repo = repositoryRoot,
): [string, SkillMetadata] {
  const root = resolvePath(path.join(repo, "skills"));
  const requested = expandUser(skill);
  let source: string;
  if (
    path.isAbsolute(requested) ||
    fs.existsSync(requested) ||
    /[/\\]/.test(skill)
  )
    source = resolvePath(requested);
  else {
    if (!namePattern.test(skill))
      throw new SkillToolError(`invalid skill name: ${skill}`);
    source = resolvePath(path.join(root, skill));
  }
  if (!isWithin(root, source))
    throw new SkillToolError(`skill source must be inside ${root}: ${source}`);
  if (!fs.existsSync(source) || !fs.statSync(source).isDirectory())
    throw new SkillToolError(
      `skill source directory does not exist: ${source}`,
    );
  return [source, requireValidSkill(source)];
}
export function resolveTarget(agent?: string, targetDir?: string): string {
  if (targetDir)
    return resolvePath(
      expandUser(
        targetDir.replace(
          /\$(\w+)|\$\{([^}]+)\}/g,
          (match, plain: string, braced: string) =>
            process.env[plain ?? braced] ?? match,
        ),
      ),
    );
  const profile = agent ?? "codex";
  let root: string;
  if (profile === "codex")
    root = process.env.CODEX_HOME ?? path.join(os.homedir(), ".codex");
  else if (profile === "pi")
    root =
      process.env.PI_CODING_AGENT_DIR ??
      path.join(os.homedir(), ".pi", "agent");
  else throw new SkillToolError(`unsupported agent profile: ${profile}`);
  return path.join(resolvePath(expandUser(root)), "skills");
}
function backupPath(root: string, name: string): string {
  const stamp = new Date()
    .toISOString()
    .replace(/[-:]/g, "")
    .replace(/\.\d{3}Z$/, "Z");
  const base = path.join(
    path.dirname(root),
    ".skill-install-backups",
    path.basename(root),
    stamp,
  );
  let candidate = path.join(base, name);
  for (let counter = 2; exists(candidate); counter++)
    candidate = path.join(base, `${name}-${counter}`);
  return candidate;
}
export interface InstallOptions {
  mode?: "link" | "copy";
  replace?: boolean;
  dryRun?: boolean;
}
export function installSkill(
  sourceInput: string,
  metadata: SkillMetadata,
  targetInput: string,
  options: InstallOptions = {},
): InstallResult {
  const source = resolvePath(sourceInput),
    target = resolvePath(expandUser(targetInput));
  const destination = path.join(target, metadata.name);
  const mode = options.mode ?? "link";
  if (destination === source)
    throw new SkillToolError(
      "installation destination is the source directory itself",
    );
  if (isWithin(source, target))
    throw new SkillToolError(
      "installation target cannot be inside the skill source",
    );
  if (!["link", "copy"].includes(mode))
    throw new SkillToolError(`unsupported install mode: ${mode}`);
  const occupied = exists(destination);
  if (
    occupied &&
    mode === "link" &&
    fs.lstatSync(destination).isSymbolicLink() &&
    resolvePath(destination) === source
  )
    return { action: "already-installed", source, destination };
  if (occupied && !options.replace)
    throw new SkillToolError(
      `destination already exists: ${destination}; rerun with --replace to back it up`,
    );
  const backup = occupied ? backupPath(target, metadata.name) : undefined;
  if (options.dryRun)
    return {
      action: occupied ? "would-replace" : "would-install",
      source,
      destination,
      backup,
    };
  fs.mkdirSync(target, { recursive: true });
  const staging = path.join(
    target,
    `.${metadata.name}.install-${randomUUID()}`,
  );
  try {
    if (mode === "link") fs.symlinkSync(source, staging, "dir");
    else {
      fs.cpSync(source, staging, {
        recursive: true,
        dereference: true,
        filter: (file) => {
          const relative = path.relative(source, file);
          if (!relative) return true;
          const name = path.basename(file);
          return (
            !ignoredDirs.has(name) &&
            name !== ".DS_Store" &&
            !/\.py[co]$/.test(name) &&
            relative !== "evals"
          );
        },
      });
      requireValidSkill(staging, metadata.name);
    }
    if (backup) {
      fs.mkdirSync(path.dirname(backup), { recursive: true });
      fs.renameSync(destination, backup);
    }
    try {
      fs.renameSync(staging, destination);
    } catch (error) {
      if (backup && exists(backup) && !exists(destination))
        fs.renameSync(backup, destination);
      throw error;
    }
  } finally {
    if (exists(staging)) fs.rmSync(staging, { recursive: true, force: true });
  }
  return {
    action: occupied ? "replaced" : "installed",
    source,
    destination,
    backup,
  };
}
export function uninstallSkill(
  name: string,
  targetInput: string,
  apply = false,
): InstallResult {
  if (!namePattern.test(name))
    throw new SkillToolError(`invalid skill name: ${name}`);
  const destination = path.join(resolvePath(expandUser(targetInput)), name);
  if (!exists(destination))
    return { action: "not-installed", source: destination, destination };
  if (!fs.lstatSync(destination).isSymbolicLink())
    requireValidSkill(destination, name);
  if (!apply)
    return { action: "would-uninstall", source: destination, destination };
  fs.rmSync(destination, { recursive: true });
  return { action: "uninstalled", source: destination, destination };
}
export function discoverSkills(repo = repositoryRoot): string[] {
  const root = path.join(repo, "skills");
  if (!fs.existsSync(root) || !fs.statSync(root).isDirectory())
    throw new SkillToolError(`skills directory does not exist: ${root}`);
  return fs
    .readdirSync(root)
    .sort()
    .map((name) => path.join(root, name))
    .filter((dir) => fs.existsSync(path.join(dir, "SKILL.md")));
}
export function runSkillTests(skill: string) {
  const dir = path.join(skill, "tests");
  const files = fs.existsSync(dir)
    ? fs
        .readdirSync(dir)
        .filter((file) => file.endsWith(".test.ts"))
        .sort()
        .map((file) => path.join(dir, file))
    : [];
  if (!files.length) return null;
  return spawnSync(
    process.execPath,
    ["--experimental-strip-types", "--test", ...files],
    { cwd: repositoryRoot, encoding: "utf8" },
  );
}
